ocaml/xcp-rrdd/bin/rrdd/xcp_rrdd.ml

```diff
@@ -545,6 +557,29 @@ let bytes_per_mem_vm = 1024
 
 let mem_vm_writer_pages = ((max_supported_vms * bytes_per_mem_vm) + 4095) / 4096
 
+let read_mem_available () =
+  let meminfo_file = "/proc/meminfo" in
```

Comment:
This could be a simpler. Below is code using `Scanf` and using the error monad (which you probably don't want and stick with exceptions). I have converted kB to bytes but maybe sticking with kB is fine, too.

```
let error fmt = Printf.kprintf (fun msg -> Result.error (`Msg msg)) fmt

let ok x = Result.ok x

let ( let* ) = Result.bind

let finally f always = Fun.protect ~finally:always f

let scanning path f =
  let io = Scanf.Scanning.open_in path in
  finally (fun () -> f io) (fun () -> Scanf.Scanning.close_in io)

let scan path =
  try scanning path @@ fun io -> Scanf.bscanf io 
    {|MemTotal: %_d %_s MemFree: %_d %_s MemAvailable: %Ld %s|}
    (fun size kb -> ok (size, kb))
  with _ -> error "failed to scan %s" path

let mem_available () =
  let* (size, kb) = scan "/proc/meminfo" in
  match kb with
  | "kB" -> ok Int64.(mul size 1024L) 
  | kb -> error "unexpected unit: %s" kb
```